modules/core/src/main/java/org/locationtech/jts/algorithm/distance/DiscreteFrechetDistance.java

```diff
@@ -0,0 +1,596 @@
+ * </pre>
+ * <p/>
+ * Its metric is better than Hausdorff's because it takes the flow of the curves
+ * into account. It is possible that two curves have a small Hausdorff but a large
+ * Fréchet distance.
+ * <p/>
+ * This implementation is base on the following optimized Fréchet distance algorithm:
+ * <pre>Thomas Devogele, Maxence Esnault, Laurent Etienne. Distance discrète de Fréchet optimisée. Spatial
+ * Analysis and Geomatics (SAGEO), Nov 2016, Nice, France. hal-02110055</pre>
+ * <p/>
+ * Several matrix storage implementations are provided
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Fr%C3%A9chet_distance">Fréchet distance</a>
+ * @see <a href="http://www.kr.tuwien.ac.at/staff/eiter/et-archive/cdtr9464.pdf">
+ *   Computing Discrete Fréchet Distance</a>
+ * @see <a href="https://hal.archives-ouvertes.fr/hal-02110055/document">Distance discrète de Fréchet optimisée</a>
+ * @see <a href="https://towardsdatascience.com/fast-discrete-fr%C3%A9chet-distance-d6b422a8fb77">
+ *   Fast Discrete Fréchet Distance</a>
+ */
+public class DiscreteFrechetDistance {
+
+  /**
+   * Computes the Discrete Fréchet Distance between two {@link Geometry}s
+   * using a {@code cartesian} distance computation function.
+   *
+   * @param g0 the 1st geometry
+   * @param g1 the 2nd geometry
+   * @return the cartesian distance between {#g0} and {#g1}
+   */
+  public static double distance(Geometry g0, Geometry g1) {
+
+    DiscreteFrechetDistance dist = new DiscreteFrechetDistance(g0, g1);
+    return dist.distance();
+  }
+
+  private final Geometry g0;
+  private final Geometry g1;
+  private PointPairDistance ptDist;
+
+  /**
+   * Creates an instance of this class using the provided geometries.
+   *
+   * @param g0 a geometry
+   * @param g1 a geometry
+   */
+  private DiscreteFrechetDistance(Geometry g0, Geometry g1) {
+    this.g0 = g0;
+    this.g1 = g1;
+  }
+
+  /**
+   * Compute the {@code Discrete Fréchet Distance} between two geometries
+   *
+   * @return the Discrete Fréchet Distance
+   */
+  private double distance() {
+    Coordinate[] coords0 = g0.getCoordinates();
+    Coordinate[] coords1 = g1.getCoordinates();
+
+    MatrixStorage distances = createMatrixStorage(coords0.length, coords1.length);
+    int[] diagonal = bresenhamLine(coords0.length, coords1.length);
+
+    HashMap<Double, int[]> distanceToPair = new HashMap<>();
+    computeCoordinateDistances(coords0, coords1, diagonal, distances, distanceToPair);
+    ptDist = computeFrechet(coords0, coords1, diagonal, distances, distanceToPair);
+
+    return ptDist.getDistance();
+  }
+
+  /**
+   * Creates a matrix storage
+   * @param rows the number of rows
+   * @param cols the number of columns
+   * @return a matrix storage
+   */
+  private MatrixStorage createMatrixStorage(int rows, int cols) {
+
+    int max = Math.max(rows, cols);
+    // NOTE: these constraints need to be verified
+    if (max < 64)
+      return new HashMapMatrix(rows, cols, Double.POSITIVE_INFINITY);
```

Comment:
Is this faster?  If not, why not just use a `RectMatrix`, since storage isn't a problem for small dimensions?